WT-13946 Set sub-level error code and message with python tests for drop EBUSY workflows

LX6T opened the pull request:
This ticket involves reproducing two of the EBUSY drop workflows and making sure that the get_last_error API works as appropriate. The idea is that the API should be able to provide more information on why an EBUSY has been returned from the drop() call. 
The following scenarios have been reproduced, handled using `WT_RET_SUB`/`WT_ERR_SUB` macros, and tested:

- `EBUSY`, `WT_CONFLICT_BACKUP`, the table is currently performing backup and cannot be dropped
- `EBUSY`, `WT_CONFLICT_DHANDLE`, another thread is currently holding the data handle of the table

jeremythorp commented on src/schema/schema_drop.c:
```diff
@@ -124,7 +131,12 @@ __drop_table(
      * table that are already open must at least be closed before this call proceeds.
      */
-    WT_ERR(__wt_schema_get_table_uri(session, uri, true, WT_DHANDLE_EXCLUSIVE, &table));
+    if ((ret = __wt_schema_get_table_uri(session, uri, true, WT_DHANDLE_EXCLUSIVE, &table)) ==
```
This code would probably be easier to read if the assignment to `ret` was pulled out of the `if`. For example, something like:
```
ret = __wt_schema_get_table_uri(session, uri, true, WT_DHANDLE_EXCLUSIVE, &table);
if (ret == EBUSY && session->err_info.err != EBUSY) {
```

LX6T replied:
Done in [Fixed catch2 drop tiered test and renamed python test file](https://github.com/wiredtiger/wiredtiger/pull/11475/commits/3235cc9bf322f4cbdb25d47828c5b5690a27e115).

jeremythorp commented on src/schema/schema_drop.c:
```diff
@@ -210,7 +222,11 @@ __drop_tiered(
     /* Get the tiered data handle. */
-    WT_RET(__wt_session_get_dhandle(session, uri, NULL, NULL, WT_DHANDLE_EXCLUSIVE));
+    if ((ret = __wt_session_get_dhandle(session, uri, NULL, NULL, WT_DHANDLE_EXCLUSIVE)) == EBUSY &&
+      session->err_info.err != EBUSY)
```
Again, pulling the assignment to `ret` out would be easier to read.

LX6T replied:
Done in [Fixed catch2 drop tiered test and renamed python test file](https://github.com/wiredtiger/wiredtiger/pull/11475/commits/3235cc9bf322f4cbdb25d47828c5b5690a27e115).

jeremythorp commented:
Thanks for creating this PR. I've partially reviewed it, and added some comments. I'll re-review once the conflicts are resolved.

LX6T commented:
> Thanks for creating this PR. I've partially reviewed it, and added some comments. I'll re-review once the conflicts are resolved.

Conflicts have been resolved and I've made some changes, ready for re-review!

DylanSHLiang commented on test/catch2/misc_tests/test_drop_conflict.cpp:
```diff
@@ -0,0 +1,100 @@
+{
+    WT_SESSION *session = NULL;
+    WT_ERROR_INFO *err_info = NULL;
+    const char *uri = "table:test_drop_conflict";
```
I think you can just put this inside the prepare_session_and_error if you are going to use the same name for all the sections.

LX6T replied:
I'm using the uri for the `open_cursor` and `drop` operations as well, so it does need to be declared in the scope of the wider test case.

LX6T replied:
I made it a macro in [Defined uri as a macro in catch2 tests](https://github.com/wiredtiger/wiredtiger/pull/11475/commits/919dbecbae6eb5726dd961adf19fddd02a8c984e), so it's no longer a parameter to `prepare_session_and_error`.

DylanSHLiang replied:
Looks good!

DylanSHLiang reviewed (approved):
LGTM! I left a comment but otherwise it looks fine.

jeremythorp commented on test/catch2/misc_tests/test_drop_conflict.cpp:
```diff
@@ -0,0 +1,101 @@
+ */
+void
+prepare_session_and_error(
+  connection_wrapper *conn_wrapper, WT_SESSION **session, WT_ERROR_INFO **err_info, std::string cfg)
```
Making `cfg` a const reference (like this `const std::string &cfg`) will avoid potentially copying the string.

LX6T replied:
Done in [Replaced reused strings with string constants, plus some minor test r…](https://github.com/wiredtiger/wiredtiger/pull/11475/commits/57343b3200b6dac9b2c6257897ab72c9bf38c1ea).

jeremythorp commented on src/schema/schema_drop.c:
```diff
@@ -27,10 +27,17 @@ __drop_file(
       session, ret = __wt_conn_dhandle_close_all(session, uri, true, force, check_visibility));
+    if (ret == EBUSY && session->err_info.err != EBUSY)
+        WT_RET_SUB(session, ret, WT_CONFLICT_DHANDLE,
+          "another thread is currently holding the data handle of the table");
```
The string `another thread is currently holding the data handle` of the table is repeated in several `WT_RET_SUB` and `WT_ERR_SUB` calls. It would be good to define that as a constant (macro) and re-use it to avoid this duplication and the risk that it gets updated in one copy later but not updated in others.

LX6T replied:
Done in [Replaced reused strings with string constants, plus some minor test r…](https://github.com/wiredtiger/wiredtiger/pull/11475/commits/57343b3200b6dac9b2c6257897ab72c9bf38c1ea).

jeremythorp commented on test/catch2/misc_tests/test_drop_conflict.cpp:
```diff
@@ -0,0 +1,101 @@
+        REQUIRE(session->open_cursor(session, URI, NULL, NULL, &cursor) == 0);
+        REQUIRE(session->drop(session, URI, NULL) == EBUSY);
+        utils::check_error_info(err_info, EBUSY, WT_CONFLICT_DHANDLE,
+          "another thread is currently holding the data handle of the table");
```
As with the code in `schema_drop.c`, the string `another thread is currently holding the data handle of the table` is repeated in several test `SECTION`s. It would be good to define that as a constant and re-use it to avoid this duplication and the risk that it gets updated in one copy later but not updated in others.

jeremythorp replied:
I suggest creating a seperate string constant in the test code (ie no need to try to reference/share the same string in the WT code) - that way if either the WT version or the test version get changed, but not the other, then tests will immediately fail.

LX6T replied:
Done in [Replaced reused strings with string constants, plus some minor test r…](https://github.com/wiredtiger/wiredtiger/pull/11475/commits/57343b3200b6dac9b2c6257897ab72c9bf38c1ea).

jeremythorp reviewed (changes requested):
Your PR is looking close now. I've made some comments, and once they're addressed it should be good to approve.

LX6T commented:
> Your PR is looking close now. I've made some comments, and once they're addressed it should be good to approve.

I've added some string constants and refactored the config string in the catch2 test, let me know if it needs anything else!

jeremythorp reviewed (approved):
Thank you, and LGTM!

jiechenbo commented on test/suite/test_error_info01.py:
```diff
@@ -26,17 +26,15 @@
 from wttest import open_cursor
 from compact_util import compact_util
 
-# test_error_info.py
```
I think we should move the EBUSY scenarios to another python file. Makes all the scenarios cleaner.

LX6T replied:
Would you mind if I left that up to the next PR? Once this gets merged I can move them all over to a new file.

jiechenbo replied:
yeah I am happy to do that.

jiechenbo reviewed (approved):
LGTM! Just one suggestion
